serverless/internal/log/integrate.go

```diff
@@ -17,25 +17,150 @@
+	}
+
+	r, err := baseRange.GetRootHash(nil)
+	if err != nil {
+		return fmt.Errorf("invalid log state, unable to recalculate root: %w", err)
+	}
+	glog.Infof("Loaded state with roothash %x", r)
+
+	tiles := make(map[string]*api.Tile)
+
+	visitor := func(id compact.NodeID, hash []byte) {
+		tileLevel, tileIndex, nodeLevel, nodeIndex := storage.NodeCoordsToTileAddress(uint64(id.Level), uint64(id.Index))
+		tileKey := tileKey(tileLevel, tileIndex)
+		tile := tiles[tileKey]
+		if tile == nil {
+			created := false
+			tile, err = st.GetTile(tileLevel, tileIndex, state.Size)
+			if err != nil {
+				if !os.IsNotExist(err) {
+					panic(err)
+				}
+				created = true
+				tile = &api.Tile{
+					Nodes: make([][]byte, 256*2),
+				}
+			}
+			glog.V(1).Infof("GetTile: %s new: %v", tileKey, created)
+			tiles[tileKey] = tile
+		}
+		tile.Nodes[api.TileNodeKey(nodeLevel, nodeIndex)] = hash
+		if nodeLevel == 0 && nodeIndex >= uint64(tile.NumLeaves) {
+			tile.NumLeaves = uint(nodeIndex + 1)
+		}
+	}
+
+	// look for new sequenced entries and build tree
+	newRange := rf.NewEmptyRange(state.Size)
+
+	// write new completed subtrees
+	n, err := st.ScanSequenced(state.Size,
+		func(seq uint64, entry []byte) error {
+			lh := h.HashLeaf(entry)
+			// Set leafhash on zeroth level
+			visitor(compact.NodeID{Level: 0, Index: seq}, lh)
+			// Update range and set internal nodes
+			newRange.Append(lh, visitor)
+			return nil
+		})
+	if err != nil {
+		return fmt.Errorf("error while integrating: %w", err)
+	}
+	if n == 0 {
+		glog.Infof("Nothing to do.")
+		// Nothing to do, nothing done.
+		return nil
+	}
+
+	if err := baseRange.AppendRange(newRange, visitor); err != nil {
+		return fmt.Errorf("failed to merge new range onto existing log: %w", err)
+	}
+
+	newRoot, err := baseRange.GetRootHash(nil)
+	if err != nil {
+		return fmt.Errorf("failed to calculate new root hash: %w", err)
+	}
+	glog.Infof("New log state: size 0x%x hash: %x", baseRange.End(), newRoot)
+
+	for k, t := range tiles {
+		l, i := splitTileKey(k)
+		if err := st.StoreTile(l, i, t); err != nil {
+			return fmt.Errorf("failed to store tile at level %d index %d: %w", l, i, err)
+		}
+	}
+
+	newState := api.LogState{
+		RootHash: newRoot,
+		Size:     baseRange.End(),
+		Hashes:   baseRange.Hashes(),
+	}
+	if err := st.UpdateState(newState); err != nil {
+		return fmt.Errorf("failed to update stored state: %w", err)
+	}
+	return nil
+}
+
+// tileKey creates a string key for the specified tile address.
+func tileKey(level, index uint64) string {
+	return fmt.Sprintf("%x/%x", level, index)
+}
+
+// splitTileKey returns the level and index implied by the given tile key.
+// This key should have been created with the tileKey function above.
+func splitTileKey(s string) (uint64, uint64) {
+	p := strings.Split(s, "/")
+	l, err := strconv.ParseUint(p[0], 16, 64)
+	if err != nil {
+		panic(err)
+	}
+	i, err := strconv.ParseUint(p[1], 16, 64)
+	if err != nil {
+		panic(err)
+	}
+	return l, i
 }
```

Comment:
Does it have to go via a string? You can just make a struct to encode the level & index, e.g. https://play.golang.org/p/E0Ef80nkSsY

Reply:
Lovely suggestion, ta.